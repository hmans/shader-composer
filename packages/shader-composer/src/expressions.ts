import { glslRepresentation } from "./glslRepresentation"

const zip = (a: TemplateStringsArray, b: any[]) => a.map((k, i) => [k, b[i]])

export type Expression = {
	_: "Expression"
	values: any[]
	render: () => string
}

export const glsl = (strings: TemplateStringsArray, ...values: any[]): Expression => ({
	_: "Expression",

	values,

	render: () =>
		zip(strings, values.map(glslRepresentation))
			.flat()
			.join("")
})

export function isExpression(v: any): v is Expression {
	return v && v._ === "Expression"
}
