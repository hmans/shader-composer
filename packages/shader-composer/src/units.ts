import {
  Camera,
  Color,
  Matrix3,
  Matrix4,
  Scene,
  Texture,
  Vector2,
  Vector3,
  Vector4,
  WebGLRenderer
} from "three"
import { Expression } from "./expressions"
import { identifier } from "./util/concatenator3000"

export type Program = "vertex" | "fragment"

/**
 * Currently supported GLSLTypes. Probably incomplete!
 */
export type GLSLType =
  | "bool"
  | "int"
  | "float"
  | "vec2"
  | "vec3"
  | "vec4"
  | "mat3"
  | "mat4"
  | "sampler2D"

export type JSTypes = {
  bool: boolean
  float: number
  int: number
  vec2: Vector2 | [number, number]
  vec3: Vector3 | Color
  vec4: Vector4
  mat3: Matrix3
  mat4: Matrix4
  sampler2D: Texture
}

export type Input<T extends GLSLType = any> = Expression | JSTypes[T] | Unit<T>

/*
`Input` used to be named `Value`. We will export an alias here for backwards compatibility
and remove it in a future release.

TODO: Remove `Value` type!
*/
export type Value<T extends GLSLType> = Input<T>

export type UpdateCallback = (
  dt: number,
  camera: Camera,
  scene: Scene,
  gl: WebGLRenderer
) => void

export type UnitConfig<T extends GLSLType> = {
  /**
   * Human-readable name of this unit.
   */
  name: string

  /**
   * Machine-readable name of the global variable for this unit.
   * Will be recreated by the compiler, so no need to set this yourself.
   */
  variableName: string

  /**
   * The GLSL type of this unit.
   */
  type: T

  /**
   * The value of this unit. Can be a reference to another unit,
   * a JavaScript type that matches this unit's GLSL type, or
   * an Expression.
   */
  value: Input<T> | undefined

  /**
   * If this is set to "vertex" or "fragment", the compiler will
   * only ever render this node in the specified program. If you
   * have units referencing gl_* variables that only exist in one
   * of the programs, use this to make sure they never appear
   * in the other program (which would lead to compilation failure.)
   */
  only?: Program

  /**
   * When set to true, this variable will automatically declare a varying,
   * calculate/source its value in the vertex program only, and pass the
   * result to the fragment program through that varying. Default: false.
   */
  varying: boolean

  /**
   * An optional uniform object. It will automatically be
   * declared in the program headers, and also made available in the
   * object returned by `compilerShader`.
   */
  uniform?: { value: JSTypes[T] }
  uniformName?: string

  /**
   * A callback that will be executed once per frame.
   */
  update?: UpdateCallback

  /**
   * A callback that will be executed when the shader is being disposed.
   */
  dispose?: () => void

  /* Chunks */
  vertex?: {
    header?: Expression
    body?: Expression
  }

  fragment?: {
    header?: Expression
    body?: Expression
  }
}

export type Unit<T extends GLSLType = any> = {
  _: "Unit"
  _unitConfig: UnitConfig<T>
}

export const Unit = <T extends GLSLType>(
  type: T,
  value: Input<T> | undefined,
  _config?: Partial<UnitConfig<T>>
) => {
  const config: UnitConfig<T> = {
    name: "Anonymous",
    type,
    value,
    varying: false,
    variableName: identifier("var", Math.floor(Math.random() * 1000000)),
    ..._config
  }

  const unit: Unit<T> = {
    _: "Unit",
    _unitConfig: config
  }

  return unit
}

export function isUnit(value: any): value is Unit {
  return value && value._ === "Unit"
}

export const isUnitInProgram = (unit: Unit, program: Program) =>
  [undefined, program].includes(unit._unitConfig.only)

export const uniformName = (unit: Unit) =>
  unit._unitConfig.uniformName ?? `u_${unit._unitConfig.variableName}`
