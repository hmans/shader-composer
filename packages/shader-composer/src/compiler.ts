import { Expression, isExpression } from "./expressions"
import { glslRepresentation } from "./glslRepresentation"
import {
	assignment,
	block,
	concatenate,
	identifier,
	Part,
	sluggify,
	statement
} from "./lib/concatenator3000"
import idGenerator from "./lib/idGenerator"
import { isUnit, isUnitInProgram, Program, Unit } from "./units"

const beginUnit = (unit: Unit) => `/*** BEGIN: ${unit._unitConfig.name} ***/`
const endUnit = (unit: Unit) => `/*** END: ${unit._unitConfig.name} ***/\n`

const collectDependencies = (
	dependencies: (Expression | Unit)[],
	state: CompilerState
) => {
	dependencies.forEach((dependency) => {
		if (isUnit(dependency)) collectUnit(dependency, state)
		else if (isExpression(dependency)) collectExpression(dependency, state)
	})
}

const collectExpression = (expression: Expression, state: CompilerState) => {
	/* Collect dependencies */
	collectDependencies(expression.values, state)
}

const collectUnitHeader = (unit: Unit, program: Program, state: CompilerState) => {
	if (unit._unitConfig[`${program}Header`])
		state[program].headers.push(
			beginUnit(unit),
			unit._unitConfig[`${program}Header`]?.render(),
			endUnit(unit)
		)
}

const collectUnitBody = (unit: Unit, program: Program, state: CompilerState) => {
	state[program].bodies.push(
		beginUnit(unit),

		/* Declare global variable */
		statement(unit.type, unit._unitConfig.variableName),

		block(
			/* Declare local value variable */
			statement(unit.type, "value", "=", glslRepresentation(unit.value)),

			/* Include body chunk, if given */
			unit._unitConfig[`${program}Body`]?.render(),

			/* Assign value back to global variable */
			assignment(unit._unitConfig.variableName, "value")
		),

		endUnit(unit)
	)
}

const collectUnit = (unit: Unit, state = CompilerState()) => {
	if (state.seen.has(unit)) return
	state.seen.add(unit)

	/* Prepare unit */
	unit._unitConfig.variableName = identifier(
		sluggify(unit._unitConfig.name),
		state.nextid()
	)

	/* Collect for vertex program */
	if (isUnitInProgram(unit, "vertex")) {
		collectDependencies(
			[unit.value, unit._unitConfig.vertexHeader, unit._unitConfig.vertexBody],
			state
		)

		collectUnitHeader(unit, "vertex", state)
		collectUnitBody(unit, "vertex", state)
	}

	/* Collect for fragment program */
	if (isUnitInProgram(unit, "fragment")) {
		collectDependencies(
			[unit.value, unit._unitConfig.fragmentHeader, unit._unitConfig.fragmentBody],
			state
		)
		collectUnitHeader(unit, "fragment", state)
		collectUnitBody(unit, "fragment", state)
	}
}

/**
 * Generates a full shader program from a compiler state object.
 */
const compileProgram = (program: Program, state: CompilerState): string =>
	concatenate(
		`/*** PROGRAM: ${program.toUpperCase()} ***/\n`,
		state[program].headers,
		"void main()",
		block(state[program].bodies)
	)

export const compileShader = (root: Unit) => {
	const state = CompilerState()
	collectUnit(root, state)

	return {
		vertexShader: compileProgram("vertex", state),
		fragmentShader: compileProgram("fragment", state)
	}
}

type CompilerState = ReturnType<typeof CompilerState>

const CompilerState = () => ({
	vertex: {
		headers: new Array<Part>(),
		bodies: new Array<Part>()
	},
	fragment: {
		headers: new Array<Part>(),
		bodies: new Array<Part>()
	},
	nextid: idGenerator(),
	seen: new Set<Unit>()
})
