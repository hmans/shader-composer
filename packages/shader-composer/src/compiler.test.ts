import { compileShader } from "./compiler"
import { glsl } from "./expressions"
import { Bool, Float, Master, Unit, Vec3 } from "./units"

describe("compileShader", () => {
	it("compiles shader programs from the given unit", () => {
		const root = Bool(true)
		const [shader] = compileShader(root)

		expect(shader.vertexShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: VERTEX ***/

		void main()
		{
		  bool Anonymous_1 = true;
		}"
	`)

		expect(shader.fragmentShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: FRAGMENT ***/

		void main()
		{
		  bool Anonymous_1 = true;
		}"
	`)
	})

	it("resolves dependencies to other units", () => {
		const f = Float(123)
		const root = Float(f)

		const [shader] = compileShader(root)

		expect(shader.vertexShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: VERTEX ***/

		void main()
		{
		  float Anonymous_2 = 123.0;
		  float Anonymous_1 = Anonymous_2;
		}"
	`)

		expect(shader.fragmentShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: FRAGMENT ***/

		void main()
		{
		  float Anonymous_2 = 123.0;
		  float Anonymous_1 = Anonymous_2;
		}"
	`)
	})

	it("resolves dependencies to other units from within expressions", () => {
		const f = Float(123)
		const root = Float(glsl`${f} * 2.0`)

		const [shader] = compileShader(root)

		expect(shader.vertexShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: VERTEX ***/

		void main()
		{
		  float Anonymous_2 = 123.0;
		  float Anonymous_1 = Anonymous_2 * 2.0;
		}"
	`)

		expect(shader.fragmentShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: FRAGMENT ***/

		void main()
		{
		  float Anonymous_2 = 123.0;
		  float Anonymous_1 = Anonymous_2 * 2.0;
		}"
	`)
	})

	it("resolves dependencies to expressions", () => {
		const a = glsl`123.0`
		const b = glsl`${a} + 4.0`
		const root = Float(glsl`${b}`)

		const [shader] = compileShader(root)

		expect(shader.vertexShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: VERTEX ***/

		void main()
		{
		  float Anonymous_1 = 123.0 + 4.0;
		}"
	`)

		expect(shader.fragmentShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: FRAGMENT ***/

		void main()
		{
		  float Anonymous_1 = 123.0 + 4.0;
		}"
	`)
	})

	it("allows units to declare themselves to exist in one program only", () => {
		const position = Float(1, { name: "Position (Vertex Only)", only: "vertex" })
		const color = Float(2, { name: "Color (Fragment Only)", only: "fragment" })

		const root = Master({
			vertex: { body: glsl`gl_Position = ${position};` },
			fragment: { body: glsl`gl_FragColor = ${color};` }
		})

		const [shader] = compileShader(root)

		expect(shader.vertexShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: VERTEX ***/

		void main()
		{
		  bool Anonymous_1;
		  {
		    bool value = true;
		    gl_Position = E_UNRESOLVED_DEPENDENCY;
		    Anonymous_1 = value;
		  }
		}"
	`)

		expect(shader.fragmentShader).toMatchInlineSnapshot(`
		"/*** PROGRAM: FRAGMENT ***/

		void main()
		{
		  bool Anonymous_1;
		  {
		    bool value = true;
		    gl_FragColor = E_UNRESOLVED_DEPENDENCY;
		    Anonymous_1 = value;
		  }
		}"
	`)
	})
})
