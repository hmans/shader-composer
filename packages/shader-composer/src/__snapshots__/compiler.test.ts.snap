// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileShader allows units to declare themselves to exist in one program only 1`] = `
"/*** PROGRAM: VERTEX ***/

void main()
{
  /*** UNIT: Position (Vertex Only) ***/
  float Position_Vertex_Only_1 = 1.0;
  
  
  /*** UNIT: Anonymous ***/
  bool Anonymous_3 = true;
  {
    bool value = Anonymous_3;
    gl_Position = Position_Vertex_Only_1;
    Anonymous_3 = value;
  }
  
  
}"
`;

exports[`compileShader allows units to declare themselves to exist in one program only 2`] = `
"/*** PROGRAM: FRAGMENT ***/

void main()
{
  /*** UNIT: Color (Fragment Only) ***/
  float Color_Fragment_Only_2 = 2.0;
  
  
  /*** UNIT: Anonymous ***/
  bool Anonymous_3 = true;
  {
    bool value = Anonymous_3;
    gl_FragColor = Color_Fragment_Only_2;
    Anonymous_3 = value;
  }
  
  
}"
`;

exports[`compileShader compiles shader programs from the given unit 1`] = `
"/*** PROGRAM: VERTEX ***/

void main()
{
  /*** UNIT: Anonymous ***/
  bool Anonymous_1 = true;
  
  
}"
`;

exports[`compileShader compiles shader programs from the given unit 2`] = `
"/*** PROGRAM: FRAGMENT ***/

void main()
{
  /*** UNIT: Anonymous ***/
  bool Anonymous_1 = true;
  
  
}"
`;

exports[`compileShader embeds body chunks in scoped blocks, with a local \`value\` variable 1`] = `
"/*** PROGRAM: FRAGMENT ***/

void main()
{
  /*** UNIT: Anonymous ***/
  bool Anonymous_1 = true;
  {
    bool value = Anonymous_1;
    gl_FragColor.rgb = vec3(1.0, 0.0, 0.0);
    Anonymous_1 = value;
  }
  
  
}"
`;

exports[`compileShader resolves dependencies to expressions 1`] = `
"/*** PROGRAM: VERTEX ***/

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0 + 4.0;
  
  
}"
`;

exports[`compileShader resolves dependencies to expressions 2`] = `
"/*** PROGRAM: FRAGMENT ***/

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0 + 4.0;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units 1`] = `
"/*** PROGRAM: VERTEX ***/

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units 2`] = `
"/*** PROGRAM: FRAGMENT ***/

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units from within expressions 1`] = `
"/*** PROGRAM: VERTEX ***/

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1 * 2.0;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units from within expressions 2`] = `
"/*** PROGRAM: FRAGMENT ***/

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1 * 2.0;
  
  
}"
`;

exports[`compileShader supports snippets 1`] = `
"/*** PROGRAM: VERTEX ***/


			float snippet_c8b24f9130(in float a) {
				return a * 2.0;
			}
void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = snippet_c8b24f9130(1.0);
  
  
}"
`;

exports[`compileShader supports snippets 2`] = `
"/*** PROGRAM: FRAGMENT ***/


			float snippet_c8b24f9130(in float a) {
				return a * 2.0;
			}
void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = snippet_c8b24f9130(1.0);
  
  
}"
`;
